contracts/cw3-fixed-multisig/src/contract.rs

```diff
@@ -0,0 +1,267 @@
+    }
+
+    // cast vote if no vote previously cast
+    ballots(&mut deps.storage, proposal_id).update(raw_sender.as_slice(), |bal| match bal {
+        Some(_) => Err(StdError::generic_err("Already voted on this proposal")),
+        None => Ok(Ballot { weight, vote }),
+    })?;
+
+    // if yes vote, update tally
+    if vote == Vote::Yes {
+        prop.yes_weight += weight;
+        proposal(&mut deps.storage).save(&proposal_id.to_be_bytes(), &prop)?;
+    }
+
+    // TODO: add event attributes
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_execute<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    proposal_id: u64,
+) -> StdResult<HandleResponse<Empty>> {
+    // anyone can trigger this if the vote passed
+
+    let mut prop = proposal_read(&deps.storage).load(&proposal_id.to_be_bytes())?;
+    if prop.status != Status::Open {
+        return Err(StdError::generic_err("Proposal is not open for voting"));
+    }
+    // we enforce this for now, but maybe it doesn't make sense... since no one can *vote*
+    // after expiration, if it is expired but there were enough yes votes, it means it passed
+    // in the period. Do we enforce execution then as well?
+    if prop.expires.is_expired(&env.block) {
+        return Err(StdError::generic_err("Proposal voting period has expired"));
+    }
+    // ensure it passed
+    if prop.yes_weight < prop.required_weight {
+        return Err(StdError::generic_err(format!(
+            "Insufficient yes votes: {} of needed {}",
+            prop.yes_weight, prop.required_weight
+        )));
+    }
+
+    // set it to executed
+    prop.status = Status::Executed;
+    proposal(&mut deps.storage).save(&proposal_id.to_be_bytes(), &prop)?;
+
+    // dispatch all proposed messages
+    // TODO: add event attributes
+    Ok(HandleResponse {
+        messages: prop.msgs,
+        log: vec![],
+        data: None,
+    })
+}
+
+pub fn handle_close<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    proposal_id: u64,
+) -> StdResult<HandleResponse<Empty>> {
+    // anyone can trigger this if the vote passed
+
+    let mut prop = proposal_read(&deps.storage).load(&proposal_id.to_be_bytes())?;
+    if [Status::Executed, Status::Failed]
+        .iter()
+        .any(|x| *x == prop.status)
+    {
+        return Err(StdError::generic_err("Cannot closed completed proposals"));
+    }
+    if !prop.expires.is_expired(&env.block) {
+        return Err(StdError::generic_err(
+            "Proposal must expire before you can close it",
+        ));
+    }
+    // ensure it did not pass (think about the above... passed and expired should be EITHER closeable or executable)
+    if prop.yes_weight >= prop.required_weight {
+        return Err(StdError::generic_err("Already passed, try to execute it"));
+    }
+
+    // set it to failed
+    prop.status = Status::Failed;
```

Comment:
This is a personal preference, but I don't like the name `Failed`, as it makes me think there was an error or something. Maybe `DidntPass`, or `Rejected`?

Also, `Executed` can be maybe be renamed to `Passed`? Though in that case I like the naming more, because it indicates the associated messages have been dispatched.

Reply:
I renamed them. Also added a pass status (for passed, but not executed). We currently do not display these all, but I will try to improve the info in query to do better.

Reply:
Updated in 025e717